server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java

```diff
@@ -607,21 +608,49 @@ void processBulkIndexIngestRequest(Task task, BulkRequest original, ActionListen
                         // (this will happen if pre-processing all items in the bulk failed)
                         actionListener.onResponse(new BulkResponse(new BulkItemResponse[0], 0));
                     } else {
-                        doExecute(task, bulkRequest, actionListener);
+                        // If a processor went async and returned a response on a different thread then
+                        // before we continue the bulk request we should fork back on a write thread:
+                        if (Thread.currentThread().getName().contains(ThreadPool.Names.WRITE)) {
```

Comment:
This comparison feels a bit fragile. Should we explicitly check if the thread changed ?

For example 
```
    void processBulkIndexIngestRequest(Task task, BulkRequest original, ActionListener<BulkResponse> listener) {
....
        threadPool.getThreadContext().putTransient("same_thread", Boolean.TRUE);
        ingestService.executeBulkRequest(
....
  } else {
                        
         Object value = threadPool.getThreadContext().getTransient("same_thread");
         if (value != null && Boolean.TRUE.equals(value)) {
              doExecute(task, bulkRequest, actionListener);
           } else {
```
Not sure which is worse the thread local or the string logic by string comparison.

Reply:
I think your idea is a little less fragile.

Reply:
I think my example may is abit incorrect, I think we would wither need to shove the current thread name into the thread local and read it back later...or just change "same_thread" to "write_thread" (with an assertion it really is the write thread) to be more correct.

Reply:
What do you think of this approach: 720ee01bf2499e4847c8afa87e7c79705624bde1

We keep track of the original thread and if we see that current thread changed compared to the original thread then we fork into a write thread, otherwise we just continue to execute the bulk.

Reply:
Yes, much better!  thanks.